File: Code Generation/responses/cpp/gemini/medium/choose-k-elements-with-maximum-sum.cpp
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

class Solution {
public:
    vector<long long> findMaxSum(vector<int>& nums1, vector<int>& nums2, int k) {
        int n = nums1.size();
        vector<long long> result(n, 0);
        
        // Store indices and sort them based on values in nums1
        vector<int> indices(n);
        for (int i = 0; i < n; ++i) {
            indices[i] = i;
        }
        
        sort(indices.begin(), indices.end(), [&](int a, int b) {
            return nums1[a] < nums1[b];
        });
        
        // Min-priority queue to keep track of the k largest elements seen so far
        priority_queue<int, vector<int>, greater<int>> pq;
        long long current_sum = 0;
        
        int j = 0;
        for (int i = 0; i < n; ++i) {
            int curr_idx = indices[i];
            
            // If the current nums1 value is the same as the previous, 
            // it will have the same sum as calculated for the first instance of this value.
            if (i > 0 && nums1[curr_idx] == nums1[indices[i - 1]]) {
                result[curr_idx] = result[indices[i - 1]];
            } else {
                // Add all elements from nums2 whose corresponding nums1 value 
                // is strictly less than the current nums1[curr_idx]
                while (j < i) {
                    int prev_idx = indices[j];
                    if (nums1[prev_idx] < nums1[curr_idx]) {
                        pq.push(nums2[prev_idx]);
                        current_sum += nums2[prev_idx];
                        
                        if (pq.size() > k) {
                            current_sum -= pq.top();
                            pq.pop();
                        }
                        j++;
                    } else {
                        break;
                    }
                }
                result[curr_idx] = current_sum;
            }
        }
        
        return result;
    }
};

File: Code Generation/responses/python/gemini/medium/choose-k-elements-with-maximum-sum.py
from typing import List
import heapq

class Solution:
    def findMaxSum(self, nums1: List[int], nums2: List[int], k: int) -> List[int]:
        n = len(nums1)
        # Create a list of tuples (nums1 value, nums2 value, original index)
        # and sort by nums1 value to process them in increasing order.
        data = sorted([(nums1[i], nums2[i], i) for i in range(n)])
        
        results = [0] * n
        min_heap = []
        current_sum = 0
        
        # We need to track the pointer to the first element in the sorted list 
        # that has the same nums1 value as the current group.
        j = 0
        for i in range(n):
            # If the current nums1 value is greater than the previous processed one,
            # we can add all values that were smaller into our potential top-k pool.
            # This handles duplicate values in nums1 correctly.
            if i > 0 and data[i][0] > data[i-1][0]:
                while j < i:
                    val_to_add = data[j][1]
                    heapq.heappush(min_heap, val_to_add)
                    current_sum += val_to_add
                    
                    # Maintain only the top k largest elements in the heap
                    if len(min_heap) > k:
                        current_sum -= heapq.heappop(min_heap)
                    j += 1
            
            # The result for the original index is the current sum of the top k
            # values from elements strictly smaller than data[i][0].
            # If nums1 values are equal, the loop won't update current_sum, 
            # fulfilling the condition nums1[j] < nums1[i].
            original_idx = data[i][2]
            results[original_idx] = current_sum
            
        return results